clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -672,6 +672,146 @@ check_function_markers()
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle nested function calls including tailcalls.
+    {
+        instr_t *label1 = INSTR_CREATE_label(GLOBAL_DCONTEXT);
+        instr_t *call1 = XINST_CREATE_call(GLOBAL_DCONTEXT, opnd_create_instr(label1));
+
+        instr_t *label2 = INSTR_CREATE_label(GLOBAL_DCONTEXT);
+        instr_t *call2 = XINST_CREATE_call(GLOBAL_DCONTEXT, opnd_create_instr(label2));
+        instr_t *ret2 = XINST_CREATE_return(GLOBAL_DCONTEXT);
+
+        instr_t *jmp = XINST_CREATE_jump(GLOBAL_DCONTEXT, opnd_create_instr(label1));
+
+        instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+        instrlist_append(ilist, call1);
+        instrlist_append(ilist, label1);
+        instrlist_append(ilist, call2);
+        instrlist_append(ilist, label2);
+        instrlist_append(ilist, ret2);
+        instrlist_append(ilist, jmp);
+
+        static constexpr addr_t BASE_ADDR = 0x123450;
+        const int call_instr_length = instr_length(GLOBAL_DCONTEXT, call1);
+        std::vector<memref_with_IR_t> memref_setup = {
+            { gen_marker(TID, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
+              nullptr },
+            { gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID), call1 },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 1), nullptr },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR,
+                         BASE_ADDR + call_instr_length),
+              nullptr },
+            { gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID), call2 },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2), nullptr },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR,
+                         BASE_ADDR + call_instr_length * 2),
+              nullptr },
+            { gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID), ret2 },
+            // A tail call to the first function.
+            { gen_instr_type(TRACE_TYPE_INSTR_DIRECT_JUMP, TID), jmp },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 1), nullptr },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR,
+                         BASE_ADDR + call_instr_length),
+              nullptr },
+        };
+        auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+        instrlist_clear_and_destroy(GLOBAL_DCONTEXT, ilist);
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle kernel transfer, sigreturn, nested function calls
+    // including tailcalls.
+    {
+        // TODO i#5949: For WOW64 instr_is_syscall() always returns false, so our
+        // checks do not currently work properly there.
+#if !defined(WINDOWS) || defined(X64)
+        // XXX: Just like raw2trace_unit_tests, we need to create a syscall instruction
+        // and it turns out there is no simple cross-platform way.
```

Comment:
I would again question why we are bothering to add encodings here: they're just adding complexity.  Are there enough sequential instructions below that manually listing the PC values is more complex?

Reply:
I've removed the encodings.